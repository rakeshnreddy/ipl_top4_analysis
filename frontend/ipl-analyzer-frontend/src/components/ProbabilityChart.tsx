import React from 'react';
import { Bar } from 'react-chartjs-2';
import { // Values from chart.js
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
  Legend
} from 'chart.js'; // Values from chart.js
import type { ChartOptions, ChartData, TooltipItem } from 'chart.js'; // Types from chart.js

import { team_short_names } from '../teamStyles'; // Removed unused team_styles and TeamStyle

ChartJS.register(
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
  Legend
);

interface ProbabilityChartProps {
  chartData: { teamKey: string; probability: number }[];
  titleText: string;
  ariaLabel: string;
}

const ProbabilityChart: React.FC<ProbabilityChartProps> = ({ chartData, titleText, ariaLabel }) => {
  const data: ChartData<'bar'> = {
    labels: chartData.map(d => team_short_names[d.teamKey] || d.teamKey),
    datasets: [
      {
        label: 'Probability',
        data: chartData.map(d => d.probability),
        // For simplicity in this subtask, using fixed colors.
        // Proper theming would require passing theme variables as props or using Chart.js scripting.
        backgroundColor: 'rgba(255, 111, 0, 0.6)', // Representative of light theme --primary-color (Vibrant Orange) with alpha
        borderColor: 'rgba(255, 111, 0, 1)',     // Opaque version for border
        borderWidth: 2,
        borderRadius: 6, // Rounded bars
        borderSkipped: false, // Border on all sides
      },
    ],
  };

  const tooltipLabelCallback = (context: TooltipItem<'bar'>): string => {
    let label = context.dataset.label || '';
    if (label) {
      label += ': ';
    }
    if (context.parsed.x !== null) {
      label += context.parsed.x.toFixed(2) + '%';
    }
    return label;
  };

  const options: ChartOptions<'bar'> = {
    indexAxis: 'y' as const,
    responsive: true,
    maintainAspectRatio: false,
    scales: {
      x: {
        beginAtZero: true,
        max: 100,
        title: {
          display: true,
          text: 'Probability (%)',
        },
        ticks: {
          callback: function(value) {
            return value + '%';
          },
          color: 'var(--text-color)' // Ensure this uses a theme-aware text color variable
        },
        grid: {
          display: false, // Remove X-axis grid lines
        }
      },
      y: {
        ticks: {
          autoSkip: false,
          color: 'var(--text-color)' // Ensure this uses a theme-aware text color variable
        },
        grid: {
          display: false, // Remove Y-axis grid lines
        },
      },
    }, // scales object ends
    plugins: {
      legend: {
        display: false,
      },
      title: {
        display: true,
        text: titleText,
        font: {
          family: "'Segoe UI', Tahoma, Geneva, Verdana, sans-serif", // Match body font
          size: 16,
        }
      },
      tooltip: {
        callbacks: {
          label: tooltipLabelCallback, // Assign the named function here
        },
      },
    },
    // Accessibility
    animation: false, // Disable animation for users sensitive to motion
    // The canvas element will get an aria-label from the Bar component props
  };

  // Height can be dynamic based on number of teams, or fixed
  const chartHeight = Math.max(300, chartData.length * 35); // Example dynamic height

  return (
    <div className="chart-container" style={{ height: `${chartHeight}px`, position: 'relative' }}>
      <Bar options={options} data={data} aria-label={ariaLabel} role="img" />
    </div>
  );
};

export default ProbabilityChart;
