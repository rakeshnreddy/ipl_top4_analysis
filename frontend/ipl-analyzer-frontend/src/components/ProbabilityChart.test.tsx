/// <reference types="vitest/globals" />
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import ProbabilityChart from './ProbabilityChart';
import type { TeamStyle } from '../teamStyles';

// Mock teamStyles to ensure consistent snapshots and avoid issues if actual styles change
vi.mock('../teamStyles', async (importOriginal) => {
  const actual = await importOriginal<typeof import('../teamStyles')>();
  return {
    ...actual,
    team_styles: {
      TeamA: { bg: '#ff0000', text: '#ffffff', accent: '#ff8800' } as TeamStyle,
      TeamB: { bg: '#00ff00', text: '#000000', accent: '#88ff00' } as TeamStyle,
    },
    team_short_names: {
      TeamA: 'TA',
      TeamB: 'TB',
    }
  };
});

// Mock ChartJS and its components as they are complex and not the focus of this unit test
vi.mock('react-chartjs-2', () => ({
  Bar: (props: any) => (
    <div data-testid="bar-chart-mock" aria-label={props['aria-label']} role={props['role']}>
      {/* You can add more details here if needed to check props passed to Bar */}
      <pre>{JSON.stringify(props.data, null, 2)}</pre>
      <pre>{JSON.stringify(props.options, null, 2)}</pre>
    </div>
  ),
}));

// Mock ChartJS registration (already done in component, but good for test isolation if needed)
vi.mock('chart.js', async (importOriginal) => {
 const actualChartJS = await importOriginal<typeof import('chart.js')>();
 return {
   ...actualChartJS,
   Chart: {
     ...actualChartJS.Chart,
     register: vi.fn(),
   },
   // Mock specific scales/elements if direct constructor calls were made in component
   // For this component, Chart.register handles it, so a simple register mock is enough.
 };
});


describe('ProbabilityChart Component', () => {
  const mockChartData = [
    { teamKey: 'TeamA', probability: 75.5 },
    { teamKey: 'TeamB', probability: 50.25 },
  ];

  const titleText = 'Test Chart Title';
  const ariaLabel = 'Test Chart Aria Label';

  it('Test 1: Renders correctly with mock data and matches snapshot', () => {
    const { container } = render(
      <ProbabilityChart chartData={mockChartData} titleText={titleText} ariaLabel={ariaLabel} />
    );

    // Check if the mock Bar chart is rendered
    expect(screen.getByTestId('bar-chart-mock')).toBeInTheDocument();
    
    // Check if title is used in options passed to Bar mock (indirectly checking title display)
    // This requires the mock to render the options.
    const optionsContent = screen.getByText(/"text": "Test Chart Title"/);
    expect(optionsContent).toBeInTheDocument();

    // Snapshot test
    // We are snapshotting the div container which includes our mock Bar chart
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Test 2: Passes correct aria-label and role to the Bar chart', () => {
    render(
      <ProbabilityChart chartData={mockChartData} titleText={titleText} ariaLabel={ariaLabel} />
    );
    const barChartMock = screen.getByTestId('bar-chart-mock');
    expect(barChartMock).toHaveAttribute('aria-label', ariaLabel);
    expect(barChartMock).toHaveAttribute('role', 'img');
  });

  it('Test 3: Calculates chart height correctly', () => {
    const { container } = render(
      <ProbabilityChart chartData={mockChartData} titleText={titleText} ariaLabel={ariaLabel} />
    );
    // chartHeight = Math.max(300, chartData.length * 35);
    // chartData.length is 2, so 2 * 35 = 70. Math.max(300, 70) = 300.
    const expectedHeight = '300px';
    // The chart container is the firstChild of the render result
    expect(container.firstChild).toHaveStyle(`height: ${expectedHeight}`);
  });
  
  it('Test 4: Calculates chart height correctly with more data', () => {
    const moreMockData = [
      ...mockChartData,
      { teamKey: 'TeamC', probability: 20 },
      { teamKey: 'TeamD', probability: 30 },
      { teamKey: 'TeamE', probability: 40 },
      { teamKey: 'TeamF', probability: 50 },
      { teamKey: 'TeamG', probability: 60 },
      { teamKey: 'TeamH', probability: 70 },
      { teamKey: 'TeamI', probability: 80 }, // 9 teams
    ];
    // chartData.length is 9, so 9 * 35 = 315. Math.max(300, 315) = 315.
    const expectedHeight = '315px';
     const { container } = render(
      <ProbabilityChart chartData={moreMockData} titleText={titleText} ariaLabel={ariaLabel} />
    );
    expect(container.firstChild).toHaveStyle(`height: ${expectedHeight}`);
  });

});
