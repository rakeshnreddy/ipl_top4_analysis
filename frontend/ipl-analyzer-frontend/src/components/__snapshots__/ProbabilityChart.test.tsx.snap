// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProbabilityChart Component > Test 1: Renders correctly with mock data and matches snapshot 1`] = `
<div
  class="chart-container"
  style="height: 300px; position: relative;"
>
  <div
    aria-label="Test Chart Aria Label"
    data-testid="bar-chart-mock"
    role="img"
  >
    <pre>
      {
  "labels": [
    "TA",
    "TB"
  ],
  "datasets": [
    {
      "label": "Probability",
      "data": [
        75.5,
        50.25
      ],
      "backgroundColor": "rgba(255, 111, 0, 0.6)",
      "borderColor": "rgba(255, 111, 0, 1)",
      "borderWidth": 2,
      "borderRadius": 6,
      "borderSkipped": false
    }
  ]
}
    </pre>
    <pre>
      {
  "indexAxis": "y",
  "responsive": true,
  "maintainAspectRatio": false,
  "scales": {
    "x": {
      "beginAtZero": true,
      "max": 100,
      "title": {
        "display": true,
        "text": "Probability (%)"
      },
      "ticks": {
        "color": "var(--text-color)"
      },
      "grid": {
        "display": false
      }
    },
    "y": {
      "ticks": {
        "autoSkip": false,
        "color": "var(--text-color)"
      },
      "grid": {
        "display": false
      }
    }
  },
  "plugins": {
    "legend": {
      "display": false
    },
    "title": {
      "display": true,
      "text": "Test Chart Title",
      "font": {
        "family": "'Segoe UI', Tahoma, Geneva, Verdana, sans-serif",
        "size": 16
      }
    },
    "tooltip": {
      "callbacks": {}
    }
  },
  "animation": false
}
    </pre>
  </div>
</div>
`;
